### PR TITLE
Signatures doc

### PR DESCRIPTION
**Pull request by kiwiroy**
### Summary
More consistency across the whole of the documentation.

### Motivation
Improve the clarity of documentation, by showing the parameters the subroutines expect on the line they are defined, rather than manipulating `@_`

\+ \> Halfway through 2019 and #1324

### References
* #1384 
* #1324

### CODE REVIEW

File: lib/Test/Mojo.pm [hunk 1]
```diff
@@ -501,8 +501,7 @@ frame type and payload.
 True if the last test was successful.
 
   # Build custom tests
-  my $location_is = sub {
-    my ($t, $value, $desc) = @_;
+  my $location_is = sub ($t, $value, $desc) {
```
Comment (Grinnz) [inline lib/Test/Mojo.pm]: $desc should be optional here

File: lib/Mojolicious/Lite.pm [hunk 2]
```diff
@@ -68,8 +68,7 @@ Mojolicious::Lite - Micro real-time web framework
   use Mojolicious::Lite;
 
   # Route with placeholder
-  get '/:foo' => sub {
-    my $c   = shift;
+  get '/:foo' => sub ($c  ) {
```
Comment (Grinnz) [inline lib/Mojolicious/Lite.pm]: whitespace mishap

File: lib/Mojolicious/Guides/Rendering.pod [hunk 3]
```diff
@@ -1516,12 +1483,11 @@ By default the renderer assumes that every C<handler> generates characters that
...
 generating bytes instead.
 
-  use Mojolicious::Lite;
+  use Mojolicious::Lite -signatures;
   use Storable 'nfreeze';
 
   # Add "storable" handler
-  app->renderer->add_handler(storable => sub {
-    my ($renderer, $c, $output, $options) = @_;
+  app->renderer->add_handler(storable => sub ($renderer, $c, $output, $options){
```
Comment (Grinnz) [inline lib/Mojolicious/Guides/Rendering.pod]: missing whitespace

**Comment by Grinnz** [comment]
Mojolicious::Guides::Testing still seems to be absent?

File: lib/Mojo/Headers.pm [hunk 4]
```diff
@@ -20,6 +20,7 @@ my %NAMES = map { lc() => $_ } (
   qw(Server-Timing Set-Cookie Status Strict-Transport-Security TE Trailer),
   qw(Transfer-Encoding Upgrade User-Agent Vary WWW-Authenticate)
 );
+
```
Comment (Grinnz) [inline lib/Mojo/Headers.pm]: Unnecessary whitespace change?

File: lib/Mojolicious/Guides/Testing.pod [hunk 5]
```diff
@@ -668,8 +667,7 @@ L<Role::Tiny> that implements a test assertion named C<location_is>:
 
   use Test::More;
 
-  sub location_is {
-    my ($t, $value, $desc) = @_;
+  sub location_is ($t, $value, $desc) {
```
Comment (Grinnz) [inline lib/Mojolicious/Guides/Testing.pod]: This $desc should also be optional

File: lib/Mojolicious/Guides/Cookbook.pod [hunk 6]
```diff
@@ -1098,7 +1068,7 @@ application.
   say 'Title: ', $res->dom->at('head > title')->text;
 
   # Extract headings
-  $res->dom('h1, h2, h3')->each(sub { say 'Heading: ', shift->all_text });
+  $res->dom('h1, h2, h3')->each(sub ($dom) { say 'Heading: ', $dom->all_text });
```
Comment (Grinnz) [inline lib/Mojolicious/Guides/Cookbook.pod]: This needs to account for the second ->each argument

File: lib/Mojolicious/Guides/Rendering.pod [hunk 7]
```diff
@@ -1108,12 +1088,10 @@ Some helpers might be useful enough for you to share them between multiple
...
 
   package Mojolicious::Plugin::DebugHelper;
-  use Mojo::Base 'Mojolicious::Plugin';
+  use Mojo::Base 'Mojolicious::Plugin', -signatures;
 
-  sub register {
-    my ($self, $app) = @_;
-    $app->helper(debug => sub {
-      my ($c, $str) = @_;
+  sub register ($self, $app) {
```
Comment (Grinnz) [inline lib/Mojolicious/Guides/Rendering.pod]: This needs to account for the third argument to register.

File: lib/Mojolicious/Guides/Rendering.pod [hunk 8]
```diff
@@ -1168,12 +1145,11 @@ plugin, and append their respective directories to the list of search paths
...
 
   package Mojolicious::Plugin::AlertAssets;
-  use Mojo::Base 'Mojolicious::Plugin';
+  use Mojo::Base 'Mojolicious::Plugin', -signatures;
 
   use Mojo::File 'path';
 
-  sub register {
-    my ($self, $app) = @_;
+  sub register ($self, $app) {
```
Comment (Grinnz) [inline lib/Mojolicious/Guides/Rendering.pod]: Also third argument to register.

File: lib/Mojolicious/Guides/Rendering.pod [hunk 9]
```diff
@@ -1209,10 +1185,9 @@ And it works just the same for assets bundled in the C<DATA> section of your
 plugin.
 
   package Mojolicious::Plugin::AlertAssets;
-  use Mojo::Base 'Mojolicious::Plugin';
+  use Mojo::Base 'Mojolicious::Plugin', -signatures;
 
-  sub register {
-    my ($self, $app) = @_;
+  sub register ($self, $app) {
```
Comment (Grinnz) [inline lib/Mojolicious/Guides/Rendering.pod]: Also third argument to register.

File: lib/Mojolicious/Guides/Rendering.pod [hunk 10]
```diff
@@ -1458,14 +1426,12 @@ for your favorite one. All you have to do, is to add a new C<handler> with
 L<Mojolicious::Renderer/"add_handler"> when C<register> is called.
 
   package Mojolicious::Plugin::MyRenderer;
-  use Mojo::Base 'Mojolicious::Plugin';
+  use Mojo::Base 'Mojolicious::Plugin', -signatures;
 
-  sub register {
-    my ($self, $app) = @_;
+  sub register ($self, $app) {
```
Comment (Grinnz) [inline lib/Mojolicious/Guides/Rendering.pod]: Also third argument to register.

File: lib/Mojolicious/Guides/Cookbook.pod [hunk 11]
```diff
@@ -1626,12 +1580,10 @@ L<Mojolicious::Plugin>. Nested helpers with a prefix based on the plugin name
...
 
   package MyApp::Plugin::MyHelpers;
-  use Mojo::Base 'Mojolicious::Plugin';
+  use Mojo::Base 'Mojolicious::Plugin', -signatures;
 
-  sub register {
-    my ($self, $app) = @_;
-    $app->helper('my_helpers.render_with_header' => sub {
-      my ($c, @args) = @_;
+  sub register ($self, $app) {
```
Comment (Grinnz) [inline lib/Mojolicious/Guides/Cookbook.pod]: Also third argument to register.

File: lib/Mojolicious/Guides/Routing.pod [hunk 12]
```diff
@@ -959,16 +932,14 @@ You can also package your conditions as reusable plugins.
...
   # Plugin
   package Mojolicious::Plugin::WerewolfCondition;
-  use Mojo::Base 'Mojolicious::Plugin';
+  use Mojo::Base 'Mojolicious::Plugin', -signatures;
 
   use Astro::MoonPhase;
 
-  sub register {
-    my ($self, $app) = @_;
+  sub register ($self, $app) {
```
Comment (Grinnz) [inline lib/Mojolicious/Guides/Routing.pod]: Also third argument to register.

File: lib/Mojolicious/Guides/Routing.pod [hunk 13]
```diff
@@ -1073,10 +1041,9 @@ You can even package applications as self-contained reusable plugins.
 
   # Plugin
   package Mojolicious::Plugin::MyEmbeddedApp;
-  use Mojo::Base 'Mojolicious::Plugin';
+  use Mojo::Base 'Mojolicious::Plugin', -signatures;
 
-  sub register {
-    my ($self, $app) = @_;
+  sub register ($self, $app) {
```
Comment (Grinnz) [inline lib/Mojolicious/Guides/Routing.pod]: Also third argument to register.

**Comment by kraih** [comment]
Where does this PR stand? How complete is it? Or has it gone stale?

**Comment by kraih** [comment]
Leaving the PR open for now with conflicts, since @kiwiroy is generally pretty active and there is probably not much work required to finish it.

File: lib/Mojo.pm [hunk 14]
```diff
@@ -42,7 +41,7 @@ Mojo - Web development toolkit
   # Event loop
   use Mojo::IOLoop;
   for my $seconds (1 .. 5) {
-    Mojo::IOLoop->timer($seconds => sub { say $seconds });
+    Mojo::IOLoop->timer($seconds => sub ($ioloop) { say $seconds });
```
Comment (kraih) [inline lib/Mojo.pm]: The invocant seems completely useless here.

Comment (kraih) [reply lib/Mojo.pm]: But we use `$loop` as a convention anyway.

**Comment by jhthorsen** [comment]
@kiwiroy: If you want this PR to move forward, then I suggest that you rebase with master, fix the conflict and call for a vote.

https://mojolicious.org/perldoc/Mojolicious/Guides/Contributing#Voting-Rules

**Comment by kiwiroy** [comment]
Rebased and fixed a few lines. Calling for a vote @mojolicious/core.

**Comment by jhthorsen** [comment]
I can't guaranty that all of the code is changed though, but I think the amount of work that has gone into this PR makes it worth merging. We can rather fix inconsistencies later on.

I'm very positive to this PR. 👍

**Comment by kiwiroy** [comment]
With 81ab932 is this worth a tidy?

**Comment by jberger** [comment]
I agree with @jhthorsen, all the more so now with the announcement of Perl 7, I want our documentation to be as modern as possible. With a rebase and a quick look for new/changed methods I will be 👍  for inclusion.

**Comment by kraih** [comment]
Yea, the time has come.

File: lib/Mojo/IOLoop.pm [hunk 15]
```diff
@@ -515,8 +499,8 @@ Accept TCP/IP and UNIX domain socket connections with L<Mojo::IOLoop::Server> an
 L<Mojo::IOLoop::Stream>, takes the same arguments as L<Mojo::IOLoop::Server/"listen">.
 
   # Listen on random port
-  my $id = Mojo::IOLoop->server({address => '127.0.0.1'} => sub {
-    my ($loop, $stream, $id) = @_;
+  my $id = Mojo::IOLoop->server({address => '127.0.0.1'} =>
+    sub ($loop, $stream, $id) {
```
Comment (Grinnz) [inline lib/Mojo/IOLoop.pm:503]: This should be collapsible to one line with the new line length limit, as well as some similar changes

File: lib/Mojo/UserAgent.pm [hunk 16]
```diff
@@ -403,27 +403,23 @@ Mojo::UserAgent - Non-blocking I/O HTTP and WebSocket user agent
...
     say $tx->result->dom->at('title')->text;
   });
   Mojo::IOLoop->start unless Mojo::IOLoop->is_running;
 
   # Concurrent non-blocking requests (synchronized with promises)
   my $mojo = $ua->get_p('mojolicious.org');
   my $cpan = $ua->get_p('cpan.org');
-  Mojo::Promise->all($mojo, $cpan)->then(sub {
-    my ($mojo, $cpan) = @_;
+  Mojo::Promise->all($mojo, $cpan)->then(sub ($mojo, $cpan) {
```
Comment (leonerd) [inline lib/Mojo/UserAgent.pm:414]: For clarity here I'd consider renaming one set of variables, to demonstrate that they're different, Maybe

```perl
my $mojo_p = $ua->get_p('mojolicious.org');
```

and then non-`_p` suffixes on the results of them. Otherwise the reader may not notice the different variables.

File: examples/chat.pl [hunk 17]
```diff
@@ -1,24 +1,26 @@
...
+  $c->on(message => sub ($c, $msg) { $c->events->emit(mojochat => $msg) });
 
   # Forward messages to the browser
   my $cb = $c->events->on(mojochat => sub { $c->send(pop) });
-  $c->on(finish => sub { shift->events->unsubscribe(mojochat => $cb) });
+  $c->on(
+    finish => sub ($c, $code, $reason = undef) {
+      $c->events->unsubscribe(mojochat => $cb);
+    }
+  );
```
Comment (kraih) [inline examples/chat.pl:23]: This should also fit on one line easily with the new line limit.

File: lib/Test/Mojo.pm [hunk 18]
```diff
@@ -506,8 +505,7 @@ Current WebSocket message represented as an array reference containing the frame
 True if the last test was successful.
 
   # Build custom tests
-  my $location_is = sub {
-    my ($t, $value, $desc) = @_;
+  my $location_is = sub ($t, $value, $desc = "") {
```
Comment (kraih) [inline lib/Test/Mojo.pm:508]: Double quotes instead of single quotes.

File: lib/Mojolicious/Validator.pm [hunk 19]
```diff
@@ -169,24 +171,28 @@ L<Mojolicious::Validator> inherits all methods from L<Mojo::Base> and implements
...
 
 Register a validation check.
 
-  $validator->add_check(foo => sub {
-    my ($v, $name, $value, @args) = @_;
-    ...
-    return undef;
-  });
+  $validator->add_check(
+    foo => sub ($v, $name, $value, @args) {
```
Comment (kraih) [inline lib/Mojolicious/Validator.pm:183]: This should be one line.

File: lib/Mojolicious/Validator.pm [hunk 20]
```diff
@@ -169,24 +171,28 @@ L<Mojolicious::Validator> inherits all methods from L<Mojo::Base> and implements
 
 =head2 add_check
 
-  $validator = $validator->add_check(size => sub {...});
+  $validator = $validator->add_check(
+    size => sub ($v, $name, $value, @args) {
```
Comment (kraih) [inline lib/Mojolicious/Validator.pm:175]: Should be one line.

File: lib/Mojolicious/Validator.pm [hunk 21]
```diff
@@ -151,7 +151,9 @@ L<Mojolicious::Validator> implements the following attributes.
 =head2 checks
 
   my $checks = $validator->checks;
-  $validator = $validator->checks({size => sub {...}});
+  $validator = $validator->checks({
+    size => sub ($v, $name, $value, @args) {...}
+  });
```
Comment (kraih) [inline lib/Mojolicious/Validator.pm:156]: Should be one line.

File: lib/Mojolicious/Routes/Route.pm [hunk 22]
```diff
@@ -611,11 +610,12 @@ Restrict HTTP methods this route is allowed to handle, defaults to no restrictio
...
-  my $route = $r->websocket('/:foo' => sub {...} => 'name');
-  my $route = $r->websocket('/:foo' => {foo => 'bar'} => sub {...});
-  my $route = $r->websocket('/:foo' => [foo => qr/\w+/] => sub {...});
-  my $route = $r->websocket('/:foo' => (agent => qr/Firefox/) => sub {...});
+  my $route = $r->websocket('/:foo' => sub ($c) {...});
+  my $route = $r->websocket('/:foo' => sub ($c) {...} => 'name');
+  my $route = $r->websocket('/:foo' => {foo => 'bar'} => sub ($c) {...});
+  my $route = $r->websocket('/:foo' => [foo => qr/\w+/] => sub ($c) {...});
+  my $route
+    = $r->websocket('/:foo' => (agent => qr/Firefox/) => sub ($c) {...});
```
Comment (kraih) [inline lib/Mojolicious/Routes/Route.pm:618]: Should be one line.

File: lib/Mojolicious/Plugin/DefaultHelpers.pm [hunk 23]
```diff
@@ -537,11 +536,12 @@ Perform non-blocking C<POST> request and forward response as efficiently as poss
 L<Mojo::UserAgent/"post"> and returns a L<Mojo::Promise> object.
 
   # Forward with exception handling
-  $c->proxy->post_p('example.com' => form => {test => 'pass'})->catch(sub {
-    my $err = shift;
-    $c->log->debug("Proxy error: $err");
-    $c->render(text => 'Something went wrong!', status => 400);
-  });
+  $c->proxy->post_p('example.com' => form => {test => 'pass'})->catch(
+    sub ($err) {
```
Comment (kraih) [inline lib/Mojolicious/Plugin/DefaultHelpers.pm:540]: Should be one line.

File: lib/Mojolicious/Lite.pm [hunk 24]
```diff
@@ -192,14 +182,15 @@ L<Mojolicious::Lite> implements the following functions, which are automatically
...
-    ['GET', 'POST'] => '/:foo' => (agent => qr/Firefox/) => sub {...};
+  my $route = any '/:foo' => sub ($c) {...};
+  my $route = any '/:foo' => sub ($c) {...} => 'name';
+  my $route = any '/:foo' => {foo => 'bar'} => sub ($c) {...};
+  my $route = any '/:foo' => [foo => qr/\w+/] => sub ($c) {...};
+  my $route = any ['GET', 'POST'] => '/:foo' => sub ($c) {...};
+  my $route =
+    any ['GET', 'POST'] => '/:foo' => [foo => qr/\w+/] => sub ($c) {...};
+  my $route = any ['GET', 'POST'] => '/:foo' => (agent => qr/Firefox/) =>
+    sub ($c) {...};
```
Comment (kraih) [inline lib/Mojolicious/Lite.pm:193]: Both examples should fit on one line.

File: lib/Mojo/UserAgent.pm (hunk 16 above)
Comment (kraih) [reply lib/Mojo/UserAgent.pm:414]: You could have said the same about the version before.

File: lib/Mojolicious/Guides/Routing.pod [hunk 25]
```diff
@@ -633,13 +620,11 @@ normal C<GET> requests with some additional information.
...
-  use Mojo::Base 'Mojolicious::Controller';
+  use Mojo::Base 'Mojolicious::Controller', -signatures;
 
   # Action
-  sub echo {
-    my $self = shift;
-    $self->on(message => sub {
-      my ($self, $msg) = @_;
+  sub echo ($self) {
+    $self->on(message => sub ($self, $msg) {
```
Comment (kraih) [inline lib/Mojolicious/Guides/Routing.pod:627]: Should be one line.

File: lib/Mojolicious/Guides/Rendering.pod [hunk 26]
```diff
@@ -1420,12 +1386,12 @@ L<Mojolicious::Renderer/"get_data_template"> to search the C<DATA> sections.
...
 
-  use Mojolicious::Lite;
+  use Mojolicious::Lite -signatures;
   use Storable qw(nfreeze);
 
   # Add "storable" handler
-  app->renderer->add_handler(storable => sub {
-    my ($renderer, $c, $output, $options) = @_;
+  app->renderer->add_handler(
+    storable => sub ($renderer, $c, $output, $options) {
```
Comment (kraih) [inline lib/Mojolicious/Guides/Rendering.pod:1394]: Should be one line.

File: lib/Mojolicious/Guides/Growing.pod [hunk 27]
```diff
@@ -243,9 +242,7 @@ concerns more visible.
 
   package MyApp::Model::Users;
 
-  use strict;
-  use warnings;
-
+  use Mojo::Base -signatures;
```
Comment (kraih) [inline lib/Mojolicious/Guides/Growing.pod:245]: This is wrong, the class is specifically meant to be non-Mojo.

File: lib/Mojolicious/Guides/Cookbook.pod [hunk 28]
```diff
@@ -1083,7 +1051,11 @@ can be a very powerful tool especially for testing web application.
   say 'Title: ', $res->dom->at('head > title')->text;
 
   # Extract headings
-  $res->dom('h1, h2, h3')->each(sub { say 'Heading: ', shift->all_text });
+  $res->dom('h1, h2, h3')->each(
+    sub ($dom, $i) {
+      say 'Heading: ', $dom->all_text;
+    }
+  );
```
Comment (kraih) [inline lib/Mojolicious/Guides/Cookbook.pod:1058]: Should fit one one line.

File: lib/Mojolicious/Guides.pod [hunk 29]
```diff
@@ -51,6 +51,12 @@ L<warnings>, L<utf8> and Perl 5.16 L<features|feature> are enabled, even if exam
 Some modules, like L<Mojo::Base> and L<Mojolicious::Lite>, will enable them for you automatically, whenever they are
 used.
 
+=item Signatures
+
+On Perl 5.20+ you can also use a C<-signatures> flag to enable support for
+L<subroutine signatures|perlsub/"Signatures">. Signatures are used in all
+examples for clarity, even when C<-signatures> is omitted for brevity.
```
Comment (kraih) [inline lib/Mojolicious/Guides.pod:58]: Needs a link to `Mojo::Base`.

File: lib/Mojo/UserAgent.pm (hunk 16 above)
Comment (kraih) [reply lib/Mojo/UserAgent.pm:414]: We don't really use `*_p` for variables, probably best to write out `$mojo_promise`.

File: lib/Mojo/Promise.pm [hunk 30]
```diff
@@ -492,26 +481,22 @@ return value of the called handler.
...
   # Change the fulfillment value or rejection reason
   $promise->then(
-    sub {
-      my @value = @_;
+    sub (@value) {
       return "This is good: $value[0]";
     },
-    sub {
-      my @reason = @_;
+    sub (@reason) {
```
Comment (kraih) [inline lib/Mojo/Promise.pm:499]: Should be one line.

Comment (kraih) [reply lib/Mojo/Promise.pm:499]: The ones above too.

File: lib/Mojo/IOLoop/Delay.pm [hunk 31]
```diff
@@ -179,24 +168,22 @@ of C<1> with no default length. The arguments are then combined in the same orde
...
     ...
   });
   Mojo::IOLoop->client({port => 3000} => $delay->begin(1, 1));
   $delay->wait;
 
   # Capture and combine arguments
-  my $delay = Mojo::IOLoop->delay(sub {
-    my ($delay, $three_err, $three_stream, $four_err, $four_stream) = @_;
+  my $delay = Mojo::IOLoop->delay(
+    sub ($delay, $three_err, $three_stream, $four_err, $four_stream) {
```
Comment (kraih) [inline lib/Mojo/IOLoop/Delay.pm:186]: Should be one line.

Comment (kraih) [reply lib/Mojo/IOLoop/Delay.pm:186]: Same for the ones above.

File: lib/Mojo/IOLoop.pm [hunk 32]
```diff
@@ -408,7 +395,8 @@ L<Mojo::IOLoop::Delay/"steps">.
   }
   my $mojo = get('https://mojolicious.org');
   my $cpan = get('https://metacpan.org');
-  Mojo::Promise->race($mojo, $cpan)->then(sub { say shift->req->url })->wait;
+  Mojo::Promise->race($mojo, $cpan)->then(sub ($tx) { say $tx->req->url })
+    ->wait;
```
Comment (kraih) [inline lib/Mojo/IOLoop.pm:399]: Should be one line.

File: examples/connect-proxy.pl [hunk 33]
```diff
@@ -43,8 +39,7 @@
                 say "Forwarding to $address:$port";
                 Mojo::IOLoop->stream($id)->write("HTTP/1.1 200 OK\x0d\x0a" . "Connection: keep-alive\x0d\x0a\x0d\x0a");
                 $stream->on(
-                  read => sub {
-                    my ($stream, $chunk) = @_;
+                  read => sub ($stream, $chunk) {
```
Comment (kraih) [inline examples/connect-proxy.pl:42]: Should fit one one line.

File: examples/connect-proxy.pl [hunk 34]
```diff
@@ -3,19 +3,15 @@
...
-  {port => 3000} => sub {
-    my ($loop, $stream, $id) = @_;
-
+  {port => 3000} => sub ($loop, $stream, $id) {
     # Connection to client
     $stream->on(
-      read => sub {
-        my ($stream, $chunk) = @_;
-
+      read => sub ($stream, $chunk) {
```
Comment (kraih) [inline examples/connect-proxy.pl:14]: Should fit on one line.

File: lib/Mojolicious/Guides/Cookbook.pod [hunk 35]
```diff
@@ -1361,8 +1326,7 @@ portable, allowing them to work inside an already running event loop or start on
   my $ua = Mojo::UserAgent->new;
   my $mojo   = $ua->get_p('https://metacpan.org/search?q=mojo');
   my $minion = $ua->get_p('https://metacpan.org/search?q=minion');
-  Mojo::Promise->all($mojo, $minion)->then(sub {
-    my ($mojo, $minion) = @_;
+  Mojo::Promise->all($mojo, $minion)->then(sub ($mojo, $minion) {
```
Comment (leonerd) [inline lib/Mojolicious/Guides/Cookbook.pod:1329]: Same thoughts here about `_p`-suffixed variables, to make clearer the distinction between those holding promises and those holding the eventual results of them

Comment (kiwiroy) [reply lib/Mojolicious/Guides/Cookbook.pod]: Missed this somehow, variables now suffixed with `_promise` as suggested elsewhere.

**Comment by kraih** [comment]
Looks like it's time for another round of reviews.

File: lib/Mojolicious/Guides/Cookbook.pod [hunk 36]
```diff
@@ -1525,15 +1482,13 @@ Fun L<Mojolicious> application hacks for all occasions.
 
 Basic authentication data will be automatically extracted from the C<Authorization> header.
 
-  use Mojolicious::Lite;
-  use Mojo::Util qw(secure_compare);
+  use Mojolicious::Lite -signatures;
+  use Mojo::Util 'secure_compare';
```
Comment (kraih) [inline lib/Mojolicious/Guides/Cookbook.pod:1486]: Accidental change from `qw(...)` to `'...'`.

File: lib/Mojolicious/Guides/Routing.pod [hunk 37]
```diff
@@ -245,15 +241,13 @@ IRIs are handled transparently, that means paths are guaranteed to be unescaped
 The generated hash of a matching route is actually the center of the whole L<Mojolicious> request cycle. We call it the
 stash, and it persists until a response has been generated.
 
-  # /bye -> {controller => 'foo', action => 'bye', mymessage => 'Bye'}
-  $r->get('/bye')
-    ->to(controller => 'foo', action => 'bye', mymessage => 'Bye');
+  # /bye          -> {controller => 'foo', action => 'bye', mymessage => 'Bye'}
```
Comment (kraih) [inline lib/Mojolicious/Guides/Routing.pod:244]: Accidental addition of whitespace.

File: lib/Mojolicious/Guides/Cookbook.pod [hunk 38]
```diff
@@ -1083,7 +1051,11 @@ can be a very powerful tool especially for testing web application.
   say 'Title: ', $res->dom->at('head > title')->text;
 
   # Extract headings
-  $res->dom('h1, h2, h3')->each(sub { say 'Heading: ', shift->all_text });
+  $res->dom('h1, h2, h3')->each(
+    sub ($dom, $i) {
```
Comment (kraih) [inline lib/Mojolicious/Guides/Cookbook.pod:1055]: This is still not on one line.
```
$res->dom('h1, h2, h3')->each(sub ($dom, $i) {
```

File: lib/Mojolicious/Guides/Growing.pod [hunk 39]
```diff
@@ -245,6 +244,8 @@ concerns more visible.
 
   use strict;
   use warnings;
+  use feature qw(signatures);
```
Comment (Grinnz) [inline lib/Mojolicious/Guides/Growing.pod:247]: This can be one line, as experimental is core in all versions that have the signatures feature: `use experimental qw(signatures);`

File: lib/Mojo/IOLoop.pm [hunk 40]
```diff
@@ -584,7 +561,9 @@ Get L<Mojo::IOLoop::Stream> object for id or turn object into a connection.
 
   my $subprocess = Mojo::IOLoop->subprocess;
   my $subprocess = $loop->subprocess;
-  my $subprocess = $loop->subprocess(sub {...}, sub {...});
+  my $subprocess = $loop->subprocess(
```
Comment (Grinnz) [inline lib/Mojo/IOLoop.pm:564]: This can probably be collapsed to one line

File: lib/Mojolicious/Guides/Cookbook.pod [hunk 41]
```diff
@@ -500,15 +496,15 @@ of concurrent client connections.
...
   # Search MetaCPAN for "mojolicious"
-  get '/' => sub {
-    my $c = shift;
-    $c->ua->get('fastapi.metacpan.org/v1/module/_search?q=mojolicious' => sub {
-      my ($ua, $tx) = @_;
-      $c->render('metacpan', hits => $tx->result->json->{hits}{hits});
-    });
+  get '/' => sub ($c) {
+    $c->ua->get(
+      'fastapi.metacpan.org/v1/module/_search?q=mojolicious' => sub ($ua, $tx) {
```
Comment (Grinnz) [inline lib/Mojolicious/Guides/Cookbook.pod:504]: This can be left on the previous line

**Comment by kraih** [comment]
Another small change is needed, conventions in the guide are no longer a list but individual headers. https://github.com/mojolicious/mojo/compare/58ea8a277bb7...0428142f1365

**Comment by marcusramberg** [comment]
🎉
